tests/test_pynml.py

```diff
@@ -0,0 +1,76 @@
+#!/usr/bin/env python3
+"""
+Unit tests for pynml.py
+
+File: test/test_pynml.py
+
+Copyright 2021 NeuroML contributors
+Author: Ankur Sinha <sanjay DOT ankur AT gmail DOT com>
+"""
+
+import unittest
+import os
+import shutil
+import logging
+
+from pyneuroml.pynml import (extract_lems_definition_files, list_exposures,
+                             list_recording_paths)
+
+
+logger = logging.getLogger(__name__)
+logger.setLevel(logging.DEBUG)
+
+
+class TestJarUtils(unittest.TestCase):
+
+    """Test jNeuroML jar related functions"""
+
+    def test_lems_def_files_extraction(self):
+        """Test extraction of NeuroML2 LEMS files from jar."""
+        filelist = ["Cells.xml",
+                    "Channels.xml",
+                    "Inputs.xml",
+                    "Networks.xml",
+                    "NeuroML2CoreTypes.xml",
+                    "NeuroMLCoreCompTypes.xml",
+                    "NeuroMLCoreDimensions.xml",
+                    "PyNN.xml",
+                    "Simulation.xml",
+                    "Synapses.xml"]
+
+        extraction_dir = extract_lems_definition_files()
+        newfilelist = os.listdir(extraction_dir)
+        shutil.rmtree(extraction_dir[:-1 * len("NeuroML2CoreTypes/")])
+        assert(sorted(filelist) == sorted(newfilelist))
+
+
+class TestHelperUtils(unittest.TestCase):
+
+    """Test helper utilities."""
+
+    def test_exposure_listing(self):
+        """Test listing of exposures in NeuroML documents."""
+        exps = list_exposures("tests/izhikevich_test_file.nml", "iz")
+        ctypes = {}
+        for key, val in exps.items():
+            ctypes[key.type] = val
+
+        self.assertTrue("izhikevich2007Cell" in ctypes.keys())
+        expnames = []
+        for exp in ctypes["izhikevich2007Cell"]:
+            expnames.append(exp.name)
+        # https://docs.neuroml.org/Userdocs/Schemas/Cells.html#schema-izhikevich2007cell
+        self.assertTrue("u" in expnames)
+        self.assertTrue("v" in expnames)
+        self.assertTrue("iMemb" in expnames)
+        self.assertTrue("iSyn" in expnames)
+
+    def test_recording_path_listing(self):
+        """Test listing of recording paths in NeuroML documents."""
+        paths = list_recording_paths("tests/izhikevich_test_file.nml", "iz")
```

Comment:
Just checking, does it list also the current from the input as recordable, e.g. https://github.com/NeuroML/NeuroML2/blob/master/LEMSexamples/LEMS_NML2_Ex2_Izh.xml#L58

Reply:
I wouldn't think so---`I` is not an exposure in the Izhikevich model:
https://docs.neuroml.org/Userdocs/Schemas/Cells.html#schema-izhikevich2007cell

`I` is an exposure in the pulsegeneratorDL component, so it'll be listed and can be recorded from there:
https://docs.neuroml.org/Userdocs/Schemas/Inputs.html?highlight=pulsegeneratordl#pulsegeneratordl

I don't quite understand why the example is recording the current for the pulsegenerator by referring to the izhikevich cell. Doesn't `izpopBurst[0]/i0/I` suggest that the pulsegenerator is a children element of the cell, which it isn't?

Reply:
Ah... that's the difference between the LEMS model as specified in the "recipe" to build the network (the LEMS file) and the built network in memory that's run. For the much of the model there is a one to one mapping between entities in the LEMS and what's in the final built model, but in actual fact, what explicitInput is saying is "put an instance of this component on the cell when you build the model", hence  izpopBurst[0]/i0/I is valid when the model is built. 

Similarly, a connection in a projection is saying "put a synapse on this cell and connect it to the event output", so the synapse component on that cell can be recorded from in the final model, see https://github.com/NeuroML/NeuroML2/blob/master/LEMSexamples/LEMS_NML2_Ex6_NMDA.xml#L89, as opposed to recording teh synapse on something like projection0/connection2/synapse1/i

In theory the list_recording_paths could be clever enough to know what elements will be in the final built model ("ah, there's an inputList, so there will be a pulse generator on cells xyz..."), but that's a bit more involved than just parsing the structure of the lems and asking what exposures they have...

Reply:
OK, I get that but then we have a few problems:

- how does one (a user or a dev) know about this? I don't think this is documented anywhere at all. Not even in the lems docs. (I mean I didn't know this until now :laughing: )
- Do all the LEMS implementations do this? (jlems/pylems/eden?): this is simple to verify.
- If they do---how is this information exposed, if it is exposed at all to the user. I.e., what should one look for? These aren't added as new exposures, so what should one look for?

The new pylems methods that get exposures do go over the full constructed model that resides in memory, and so do these helper functions. We're not text parsing the LEMS file and figuring these things out---we're constructing the full model in PyLEMS and then getting exposures from the constructed model. So if this information is exposed, we can add helper methods to make them available to users too.